app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/CachedTimelineRemoteMediator.kt

```diff
@@ -0,0 +1,154 @@
+                        }
+                    }
+                }
+                initialRefresh = true
+                dbEmpty = topId == null
+            }
+
+            val statusResponse = when (loadType) {
+                LoadType.REFRESH -> {
+                    api.homeTimeline(limit = state.config.pageSize).await()
+                }
+                LoadType.PREPEND -> {
+                    return MediatorResult.Success(endOfPaginationReached = true)
+                }
+                LoadType.APPEND -> {
+                    val maxId = state.pages.findLast { it.data.isNotEmpty() }?.data?.lastOrNull()?.status?.serverId
+                    api.homeTimeline(maxId = maxId, limit = state.config.pageSize).await()
+                }
+            }
+
+            val statuses = statusResponse.body()
+            if (!statusResponse.isSuccessful || statuses == null) {
+                return MediatorResult.Error(HttpException(statusResponse))
+            }
+
+            db.withTransaction {
+                val overlappedStatuses = replaceStatusRange(statuses, state)
+
+                if (loadType == LoadType.REFRESH && overlappedStatuses == 0 && statuses.isNotEmpty() && !dbEmpty) {
+                    timelineDao.insertStatus(
+                        Placeholder(statuses.last().id.dec(), loading = false).toEntity(activeAccount.id)
+                    )
+                }
+            }
+            return MediatorResult.Success(endOfPaginationReached = statuses.isEmpty())
+        } catch (e: Exception) {
+            return MediatorResult.Error(e)
+        }
+    }
+
+    /**
+     * Deletes all statuses in a given range and inserts new statuses.
+     * This is necessary so statuses that have been deleted on the server are cleaned up.
+     * Should be run in a transaction as it executes multiple db updates
+     * @param statuses the new statuses
+     * @return the number of old statuses that have been cleared from the database
+     */
+    private suspend fun replaceStatusRange(statuses: List<Status>, state: PagingState<Int, TimelineStatusWithAccount>): Int {
+        val overlappedStatuses = if (statuses.isNotEmpty()) {
+            timelineDao.deleteRange(activeAccount.id, statuses.last().id, statuses.first().id)
+        } else {
+            0
+        }
+
+        for (status in statuses) {
+            timelineDao.insertAccount(status.account.toEntity(activeAccount.id, gson))
+            status.reblog?.account?.toEntity(activeAccount.id, gson)?.let { rebloggedAccount ->
+                timelineDao.insertAccount(rebloggedAccount)
+            }
+
+            // check if we already have one of the newly loaded statuses cached locally
+            // in case we do, copy the local state (expanded, contentShowing, contentCollapsed) over so it doesn't get lost
+            var oldStatus: TimelineStatusEntity? = null
+            for (page in state.pages) {
+                oldStatus = page.data.find { s ->
```

Comment:
This is a triple-nested loop which makes me a bit uneasy but I guess pages don't overlap so it is like 2 loops. Still not ideal but will do.

Reply:
yeah it is not pretty :/